cmd/stellar-rpc/lib/preflight/src/lib.rs

```diff
@@ -35,23 +35,84 @@ extern crate soroban_simulation_prev;
 #[path = "."]
 mod curr {
     pub(crate) use soroban_env_host_curr as soroban_env_host;
+    pub(crate) use soroban_env_host_curr::xdr;
     pub(crate) use soroban_simulation_curr as soroban_simulation;
+
     #[allow(clippy::duplicate_mod)]
     pub(crate) mod shared;
 
     pub(crate) const PROTOCOL: u32 = soroban_env_host::meta::INTERFACE_VERSION.protocol;
+
+    use std::{rc::Rc, result::Result};
+
+    // Protocol 23 dropped the SnapshotSourceWithArchive trait in lieu of just
+    // SnapshotSource. This means our GoLedgerStorage structure needs to
+    // implement different traits (get vs. get_including_archived, for Protocol
+    // 23 and 22, respectively)
+    impl soroban_env_host::storage::SnapshotSource for crate::GoLedgerStorage {
+        fn get(
+            &self,
+            key: &Rc<xdr::LedgerKey>,
+        ) -> Result<
+            Option<soroban_env_host::storage::EntryWithLiveUntil>,
+            soroban_env_host::HostError,
+        > {
+            use xdr::{ScErrorCode, ScErrorType};
+
+            match shared::get_fallible_from_go_ledger_storage(self, key.as_ref()) {
+                Ok(res) => Ok(res),
+                Err(e) => {
+                    // Store the internal error in the storage as the info won't
+                    // be propagated from simulation.
+                    if let Ok(mut err) = self.internal_error.try_borrow_mut() {
+                        *err = Some(e);
+                    }
+                    // Errors that occur in storage are not recoverable, so we
+                    // force host to halt by passing it an internal error.
+                    Err((ScErrorType::Storage, ScErrorCode::InternalError).into())
+                }
+            }
```

Comment:
Can you factor this out? (since it's identical in for `soroban_simulation::SnapshotSourceWithArchive` as well)

Reply:
I thought I could, but it doesn't work because `ScErrorType` and `ScErrorCode` come from different XDR libraries (`prev::xdr` and `curr::xdr`) for the two implementations :cry:

Reply:
jk I figured it out :sunglasses: see 9492eb4

Reply:
Nice!